core/src/main/scala/kafka/server/ClientMetricsManager.scala

```diff
@@ -62,11 +108,92 @@ class ClientMetricsManager {
     }
 
-    clientInstance.updateLastAccessTs(Calendar.getInstance.getTime.getTime)
+    clientInstance.updateLastAccessTs(getCurrentTime)
 
     new GetTelemetrySubscriptionResponse(data)
   }
 
+  def validatePushRequest(pushTelemetryRequest: PushTelemetryRequest,
+                          clientInfo: CmClientInformation): Unit = {
+
+    def isSupportedCompressionType(id: Int) : Boolean = {
+      try {
+        CompressionType.forId(id)
+        true
+      } catch {
+        case e: IllegalArgumentException => false
+      }
+    }
+
+    val clientInstanceId = Option(pushTelemetryRequest.getClientInstanceId) match {
+      case None | Some(Uuid.ZERO_UUID) =>
+        val msg = String.format("Invalid request from the client [%s], missing client instance id", clientInfo.getClientId)
+        throw new ClientMetricsException(msg, Errors.INVALID_REQUEST)
+      case _ => pushTelemetryRequest.getClientInstanceId
+    }
+
+    val clientInstance = getClientInstance(clientInstanceId).getOrElse(createClientInstance(clientInstanceId, clientInfo))
+    if (pushTelemetryRequest.getSubscriptionId != clientInstance.getSubscriptionId) {
+      val msg = String.format("Client's subscription id [%s] != Broker's cached client's subscription id [%s]",
+        pushTelemetryRequest.getSubscriptionId.toString, clientInstance.getSubscriptionId.toString)
+      throw new ClientMetricsException(msg, Errors.UNKNOWN_CLIENT_METRICS_SUBSCRIPTION_ID)
+    }
+
+    if (!isSupportedCompressionType(pushTelemetryRequest.data().compressionType)) {
+      val msg = String.format("Unknown compression type [%s] is received in PushTelemetryRequest from %s",
+        pushTelemetryRequest.data().compressionType().toString, pushTelemetryRequest.getClientInstanceId.toString)
+      throw new ClientMetricsException(msg, Errors.UNSUPPORTED_COMPRESSION_TYPE)
+    }
+
+    if (!clientInstance.canAcceptPushRequest(pushTelemetryRequest.isClientTerminating)) {
+      val msg = String.format("Request from the client [%s] arrived before the throttling time",
+        pushTelemetryRequest.getClientInstanceId.toString)
+      throw new ClientMetricsException(msg, Errors.THROTTLING_QUOTA_EXCEEDED)
+    }
+  }
+
+  def processPushTelemetryRequest(pushTelemetryRequest: PushTelemetryRequest,
+                                  requestContext: RequestContext,
+                                  clientInfo: CmClientInformation,
+                                  throttleMs: Int): PushTelemetryResponse = {
+
+    def createResponse(errors: Option[Errors]) : PushTelemetryResponse = {
+      var adjustedThrottleMs = throttleMs
+      val clientInstance = getClientInstance(pushTelemetryRequest.getClientInstanceId)
+
+      // Before sending the response make sure to update the book keeping markers like
+      // lastAccessTime, isTerminating flag etc..
+      if (!clientInstance.isEmpty) {
+        adjustedThrottleMs = Math.max(clientInstance.get.getAdjustedPushInterval(), throttleMs)
```

Comment:
I'm still not super fond of calling this throttling since it's ambiguous, particularily in the request/response context.
And we're not actually throttling these requests, we're dropping them, so it is more of a rate limiter.

Reply:
Agreed but the pushTelemetryRequest RPC message parameter still says it is throttleMs.. I guess if we have to change we need to change in both the places.